src/ucp/rndv/rndv.c

```diff
@@ -848,25 +869,26 @@ ucp_rndv_send_frag_get_mem_type(ucp_request_t *sreq, ucs_ptr_map_key_t rreq_id,
         ucs_fatal("failed to allocate fragment memory desc");
     }
 
-    freq->send.ep = sreq->send.ep;
+    freq->send.ep            = sreq->send.ep;
+    freq->send.rndv_get.rkey = rkey;
```

Comment:
@bureddy seems we are passing rkey for inter-process transport to a request which is supposed to only copy between memtypes? so it should be NULL here?
@souravzzz I think we should move the code from ucp_rndv_init_mem_type_frag_req() into this function to replace lines 865-869 (in the version of the code "before" this PR

Reply:
@yosefe this func will used in 2 cases 
(1) to copy from between mem type from Deviece to HOST  in  put pipeline. in this case rkey is NULL for CUDA 
(2) to copy from remote memory to local fragment in get pipeline (for example in H D case). in this case we have to pass the inter process transport rkey.

Reply:
Hi @yosefe To clarify the the change you suggested, do you mean replacing the 

```
for (i = 0; i < UCP_MAX_LANES; i++) {
    freq->send.rndv_get.rkey_index[i] = rkey_index ? rkey_index[i]
                                                       : UCP_NULL_RESOURCE;
}
```
with something like

```
if (rkey_index != NULL) {
    memcpy(freq->send.rndv_get.rkey_index, rkey_index, UCP_MAX_LANES * sizeof(uint8_t));
} else {
    memset(freq->send.rndv_get.rkey_index, UCP_NULL_RESOURCE, UCP_MAX_LANES * sizeof(uint8_t));
}
```

Reply:
@souravzzz can we do the rkey pack/unpack logic (in a separate) function, before calling ucp_rndv_send_frag_get_mem_type() from ucp_rndv_send_start_put_pipeline()? IIUC we need the rkey-of-mdesc only for the case (1) above [copy D->H in put pipeline]. this way no return  the *rkey_p in a weird way from ucp_rndv_init_mem_type_frag_req()

Reply:
@yosefe I have added a function `ucp_rndv_send_frag_update_get_rkey()` that does this when `UCP_MEM_IS_GPU(remote_mem_type)` holds true to further simplify the flow. Does it look ok now?

Reply:
@souravzzz AFAIU, ucp_rndv_send_frag_get_mem_type() is used for both case (1) and (2) but we need to set rkey from mdesc only for case(1). so maybe move the usage of new function to ucp_rndv_send_start_put_pipeline() ? then only case 1 will be affected

Reply:
@yosefe I tried to isolate the change to ucp_rndv_send_start_put_pipeline() but could not do so because then the functionality of ucp_rndv_send_frag_get_mem_type() will need to be split into two parts: 1) first half where we get the mdsc by calling ucp_worker_mpool_get(), and 2) where we call ucp_request_send() to start the GET. The ucp_rndv_send_frag_update_get_rkey() will have to go in between these 2 functions. This seems to complicate the flow even more. Thoughts?

Reply:
@souravzzz can we pass param (1/0) from ucp_rndv_send_start_put_pipeline to ucp_rndv_send_frag_get_mem_type which controls whether to call ucp_rndv_send_frag_update_get_rkey? instead of doing ` if (UCP_MEM_IS_GPU(remote_mem_type)) {`

Reply:
to clarify, this is instead of breaking down ucp_rndv_send_frag_get_mem_type to 2 parts

Reply:
@yosefe Let me know if 724732cbf230ab049f94efd59a6826dec1651f2c looks good